anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/CusumAnomalyDetector.java

```diff
@@ -0,0 +1,257 @@
+        final double stdDev = sqrt(variance);
+        final double slackValue = 0.5 * stdDev;
+        final double weakThreshold = weakThresholdSigmas * stdDev;
+        final double strongThreshold = strongThresholdSigmas * stdDev;
+
+        this.highCusum = Math.max(0, highCusum + observed - (targetValue + slackValue));
+        this.lowCusum = Math.min(0, lowCusum + observed - (targetValue - slackValue));
+
+        Double weakThresholdUpper = null;
+        Double weakThresholdLower = null;
+        Double strongThresholdUpper = null;
+        Double strongThresholdLower = null;
+
+        AnomalyLevel anomalyLevel = NORMAL;
+        switch (tail) {
+        case LEFT_TAILED:
+            weakThresholdLower = -weakThreshold;
+            strongThresholdLower = -strongThreshold;
+            if (lowCusum <= strongThresholdLower) {
+                anomalyLevel = STRONG;
+            } else if (lowCusum <= weakThresholdLower) {
+                anomalyLevel = WEAK;
+            }
+            break;
+        case RIGHT_TAILED:
+            weakThresholdUpper = weakThreshold;
+            strongThresholdUpper = strongThreshold;
+            if (highCusum >= strongThresholdUpper) {
+                anomalyLevel = STRONG;
+            } else if (highCusum > weakThresholdUpper) {
+                anomalyLevel = WEAK;
+            }
+            break;
+        case TWO_TAILED:
+            weakThresholdLower = -weakThreshold;
+            strongThresholdLower = -strongThreshold;
+            weakThresholdUpper = weakThreshold;
+            strongThresholdUpper = strongThreshold;
+            if (highCusum >= strongThreshold || lowCusum <= strongThreshold) {
+                anomalyLevel = STRONG;
+            } else if (highCusum > weakThreshold || lowCusum <= weakThreshold) {
+                anomalyLevel = WEAK;
+            }
+            break;
+        default:
+            throw new IllegalStateException("Illegal tail: " + tail);
+        }
+
+        final Mpoint mpoint = MetricPointUtil.toMpoint(metricPoint);
+        final AnomalyResult result = new AnomalyResult();
+        result.setMetric(mpoint.getMetric());
+        result.setDetectorId(this.getId());
+        result.setEpochSecond(mpoint.getInstant().getEpochSecond());
+        result.setObserved(observed);
+        result.setPredicted(targetValue);
+        result.setWeakThresholdUpper(weakThresholdUpper);
+        result.setWeakThresholdLower(weakThresholdLower);
+        result.setStrongThresholdUpper(strongThresholdUpper);
+        result.setStrongThresholdLower(strongThresholdLower);
+        result.setAnomalyScore(dist);
+        result.setAnomalyLevel(anomalyLevel);
+
+        updateMeanAndVariance(observed);
+
+        return result;
+    }
+
+    private void setTargetValue(double mean) {
+        if (targetValue == 0.0) {
+            this.targetValue = mean;
+        }
+    }
+
+    private void updateMeanAndVariance(double value) {
```

Comment:
This doesn't seem to belong here. CUSUM doesn't have a mean and variance estimate. There should be a k for the slack and then the SH/SL buckets.